devtools/common/src/main/java/io/quarkus/cli/commands/AddExtensions.java

```diff
@@ -31,57 +33,96 @@ public AddExtensions(final ProjectWriter writer, final String pom) throws IOExce
+        boolean success = true;
         List<Dependency> dependenciesFromBom = getDependenciesFromBom();
 
-        for (String dependency : extensions) {
-            List<Extension> matches = MojoUtils.loadExtensions().stream()
-                    .filter(d -> {
-                        boolean hasTag = d.labels().contains(dependency.trim().toLowerCase());
-                        boolean machName = d.getName()
-                                .toLowerCase()
-                                .contains(dependency.trim().toLowerCase());
-                        boolean matchArtifactId = d.getArtifactId()
-                                .toLowerCase()
-                                .contains(dependency.trim().toLowerCase());
-                        return hasTag || machName || matchArtifactId;
-                    })
-                    .collect(Collectors.toList());
-
-            if (matches.size() > 1) {
+        for (String query : extensions) {
+            List<Extension> registry = MojoUtils.loadExtensions();
+
+            SelectionResult result = select(query, registry);
+
+            if (!result.matches()) {
                 StringBuilder sb = new StringBuilder();
-                sb.append(NOK)
-                        .append(" Multiple extensions matching '" + dependency + "'");
-
-                matches.stream()
-                        .forEach(extension -> sb.append(System.lineSeparator()).append("     * ")
-                                .append(extension.managementKey()));
-                sb.append(System.lineSeparator()).append("     Be more specific e.g using the exact name or the full gav.");
-                System.out.println(sb);
-            } else if (matches.size() == 1) {
-                final Extension extension = matches.get(0);
-
-                if (!MojoUtils.hasDependency(model, extension.getGroupId(), extension.getArtifactId())) {
-                    System.out.println(OK + " Adding extension " + extension.managementKey());
-                    model.addDependency(extension
-                            .toDependency(containsBOM(model) &&
-                                    isDefinedInBom(dependenciesFromBom, extension)));
-                    updated = true;
+                // We have 3 cases, we can still have a single candidate, but the match is on label
+                // or we have several candidates, or none
+                Set<Extension> candidates = result.getExtensions();
+                if (candidates.isEmpty() && query.contains(":")) {
+                    updated = addExtensionAsGAV(query) || updated;
+                } else if (candidates.isEmpty()) {
+                    // No matches at all.
+                    print(NOK + " Cannot find a dependency matching '" + query + "', maybe a typo?");
+                    success = false;
+                } else if (candidates.size() == 1) {
+                    sb.append(NOK).append(" One extension matching '").append(query).append("'");
```

Comment:
I don't understand if there is a single matching element, why do you NOK it?

Reply:
because it's typically a single match on labels. (you wrote the proposal ;-))

Reply:
Should we change this? 

Imagine that there is a single match but it comes from labels. Should we ask the user to be more explicit?

Reply:
Maybe I don’t understand when this case is happening then. 
I think in the proposal I proposed to not search on labels at all. Here you search but complain to the user? Or is it something else ?

Reply:
You are right it still looks for labels and _indicate_ to the user the matching extensions. 
I will remove the label lookup.

Reply:
Removed.

Reply:
Hey @cescoffier I've been thinking about it some more.
I now (aka today ;) ) that removing labels might be a bit aggressive. For example if one wants to find smallrye-openapi when one types Swagger, the  way to do it is to have it in the name. Which is constraining. So Maybe a label search make sense afterall.
Since they are treated less importantly than shortname and "full" name match, this might be acceptable.

Reply:
The one option we have is to auto accept them, or do like you did  here and stop the algorithm by asking to select/confirm the extension explicitly.
I was fine with option 1 but I can be convinced otherwise. If we go with option 2, we need a clearer text to describe what is going on as the general non matching text is confusing.